File: 조준장/level2/거리두기 확인하기.py
def bfs(graph):
    dx = [1, -1, 0, 0, 2, -2, 0, 0, -1, -1, 1, 1]
    dy = [0, 0, 1, -1, 0, 0, 2, -2, 1, -1, 1, -1]
    for x in range(5):
        for y in range(5):
            if graph[x][y] == "P":
    
                for i in range(12): # 아래/위/오/왼/아래2/위2/오2/왼2/위우대각선/위왼대각선/아래우대각선/아래왼대각선
                    nx = x + dx[i]
                    ny = y + dy[i]

                    if 0 <= nx < 5 and 0 <= ny < 5:
                        if i in [0, 1, 2, 3]:
                            if graph[nx][ny] == "P":
                                return 0
                        
                        if i == 4:
                            if graph[nx][ny] == "P" and graph[nx-1][ny] == "O":
                                return 0
                            
                        if i == 5:
                            if graph[nx][ny] == "P" and graph[nx+1][ny] == "O":
                                return 0
                        
                        if i == 6:
                            if graph[nx][ny] == "P" and graph[nx][ny-1] == "O":
                                return 0
                        
                        if i == 7:
                            if graph[nx][ny] == "P" and graph[nx][ny+1] == "O":
                                return 0

                        if i == 8:
                            if graph[nx][ny] == "P" and (graph[x-1][y] != "X" or graph[x][y+1] != "X"):
                                return 0

                        if i == 9:
                            if graph[nx][ny] == "P" and (graph[x-1][y] != "X" or graph[x][y-1] != "X"):
                                return 0

                        if i == 10:
                            if graph[nx][ny] == "P" and (graph[x+1][y] != "X" or graph[x][y+1] != "X"):
                                return 0

                        if i == 11:
                            if graph[nx][ny] == "P" and (graph[x+1][y] != "X" or graph[x][y-1] != "X"):
                                return 0
                        
    return 1


def solution(places):
    return [bfs(place) for place in places]

File: 이훈일/Programmers/Level2/거리두기 확인하기.java
import java.util.*;

class Solution {
    int[] dx = {-1, 1, 0, 0};
    int[] dy = {0, 0, -1, 1};
    
    public int[] solution(String[][] places) {
        int[] answer = new int[5];
        
        for(int i = 0; i < 5; i++) {
            String[] graph = places[i];
            boolean isPossible = true;           
            for(int j = 0; j < 5 && isPossible; j++) {
                for(int k = 0; k < 5 && isPossible; k++) {
                    if(graph[j].charAt(k) == 'P') {
                        if(!bfs(j, k, graph)) {
                            isPossible = false;
                        }
                    }
                }
            }
            
            answer[i] = isPossible ? 1 : 0;
        }
        
        return answer;
    }
    
    public boolean bfs(int row, int col, String[] graph) {
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[]{row, col});
        
        while(!queue.isEmpty()) {
            int[] node = queue.poll();
            int x = node[0];
            int y = node[1];
            
            for(int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];

                if(nx < 0 || nx >= 5 || ny < 0 || ny >= 5 || (nx == row && ny == col)) {
                    continue;
                }
                
                int d = Math.abs(nx - row) + Math.abs(ny - col);
                    
                if(graph[nx].charAt(ny) == 'P' && d <= 2) {
                    return false;
                } else if(graph[nx].charAt(ny) == 'O' && d < 2) {
                    queue.add(new int[]{nx, ny});
                }
            }
        }
        
        return true;
    }
}
